### PR TITLE
support proxylessnas with NNI NAS APIs

### PR DESCRIPTION
**Pull request by QuanluZhang**
The current version does not involve flops estimation, and only support gradient descent for architecture weights.

TODO: documentation and experiment result reproduce.

### CODE REVIEW

File: examples/nas/proxylessnas/datasets.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+# Copyright (c) Microsoft Corporation
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/datasets.py]: seems we don't use this long format anymore?

File: examples/nas/proxylessnas/datasets.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+# All rights reserved.
+#
+# MIT License
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/datasets.py]: and seems we don't add the license in example?

File: examples/nas/proxylessnas/datasets.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    def train_path(self):
+        return os.path.join(self.save_path, 'train')
+
+    @property
+    def valid_path(self):
+        return os.path.join(self._save_path, 'val')
+
+    @property
+    def normalize(self):
+        return transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/datasets.py]: there is too much code about image data processing, can we consider merge them as one?

File: examples/nas/proxylessnas/main.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    # configurations of the model
+    parser.add_argument("--n_cell_stages", default='4,4,4,4,4,1', type=str)
+    parser.add_argument("--stride_stages", default='2,2,2,1,2,1', type=str)
+    parser.add_argument("--width_stages", default='24,40,80,96,192,320', type=str)
+    parser.add_argument("--bn_momentum", default=0.1, type=float)
+    parser.add_argument("--bn_eps", default=1e-3, type=float)
+    parser.add_argument("--dropout_rate", default=0, type=float)
+    parser.add_argument("--no_decay_keys", default='bn', type=str, choices=[None, 'bn', 'bn#bias'])
+    # configurations of imagenet dataset
+    parser.add_argument("--data_path", default='/data/ssd1/v-yugzh/imagenet/', type=str)
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/main.py]: change the default data path?

File: examples/nas/proxylessnas/main.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    #parser.add_argument("--data_path", default='/mnt/v-yugzh/imagenet/', type=str)
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/main.py]: remove the comments?

File: examples/nas/proxylessnas/model.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/model.py]: license format?

File: examples/nas/proxylessnas/model.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+        input_channel = putils.make_divisible(32 * width_mult, 8)
+        first_cell_width = putils.make_divisible(16 * width_mult, 8)
+        for i in range(len(width_stages)):
+            width_stages[i] = putils.make_divisible(width_stages[i] * width_mult, 8)
+        # first conv
+        first_conv = ops.ConvLayer(3, input_channel, kernel_size=3, stride=2, use_bn=True, act_func='relu6', ops_order='weight_bn_act')
+        # first block
+        first_block_conv = ops.OPS['3x3_MBConv1'](input_channel, first_cell_width, 1)
+        #first_block = ops.MobileInvertedResidualBlock(first_block_conv, None, False)
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/model.py]: remove the comments?

File: examples/nas/proxylessnas/ops.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
+# Copyright (c) Microsoft Corporation
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/ops.py]: same here.

File: examples/nas/proxylessnas/ops.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+    def __init__(self, groups):
+        super(ShuffleLayer, self).__init__()
+        self.groups = groups
+
+    def forward(self, x):
+        batchsize, num_channels, height, width = x.size()
+        channels_per_group = num_channels // self.groups
+        # reshape
+        x = x.view(batchsize, self.groups, channels_per_group, height, width)
+        # noinspection PyUnresolvedReferences
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/ops.py]: ???

File: examples/nas/proxylessnas/ops.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        x = torch.transpose(x, 1, 2).contiguous()
+        # flatten
+        x = x.view(batchsize, -1, height, width)
+        return x
+
+class My2DLayer(nn.Module):
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/ops.py]: the name is wired... `2DLayer` is OK?

File: examples/nas/proxylessnas/ops.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+            x = module(x)
+        return x
+
+    @staticmethod
+    def is_zero_layer():
+        return False
+
+
+class MBInvertedConvLayer(nn.Module):
+
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/ops.py]: add some reference here?

File: examples/nas/proxylessnas/putils.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
+# Copyright (c) Microsoft Corporation
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/putils.py]: same here.

File: examples/nas/proxylessnas/putils.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        raise ValueError('do not support: %s' % act_func)
+
+
+def make_divisible(v, divisor, min_val=None):
+    """
+    This function is taken from the original tf repo.
+    It ensures that all layers have a channel number that is divisible by 8
+    It can be seen here:
+    https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet.py
+    :param v:
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/putils.py]: docstring format?

File: examples/nas/proxylessnas/putils.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    new_v = max(min_val, int(v + divisor / 2) // divisor * divisor)
+    # Make sure that round down does not go down by more than 10%.
+    if new_v < 0.9 * v:
+        new_v += divisor
+    return new_v
+
+class AverageMeter(object):
+    """
+    Computes and stores the average and current value
+    Copied from: https://github.com/pytorch/examples/blob/master/imagenet/main.py
```
Comment (xuehui1991) [inline examples/nas/proxylessnas/putils.py]: ???

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 15]
```diff
@@ -77,6 +77,6 @@ def apply_fixed_architecture(model, fixed_arc_path, device=None):
             fixed_arc = json.load(f)
     fixed_arc = _encode_tensor(fixed_arc, device)
     architecture = FixedArchitecture(model, fixed_arc)
-    architecture.to(device)
+    #architecture.to(device)
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py]: remove

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 16]
```diff
@@ -0,0 +1,471 @@
...
+
+        return grad_x[0], binary_grads, None, None
+
+class MixedOp(nn.Module):
+    """
+    This class is to instantiate and manage info of one LayerChoice.
+    It includes architecture weights, binary weights, and member functions
+    operating the weights.
+    """
+    # forward/backward mode for LayerChoice: None, two, full, and full_v2.
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: put into docstring

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        Parameters
+        ----------
+        mutable : LayerChoice
+            this layer's mutable
+        """
+        self.log_prob = None
+        # reset binary gates
+        self.AP_path_wb.data.zero_()
+        probs = self.probs_over_ops
+        if MixedOp.forward_mode == 'two':
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: ???? poor code readability

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/trainer.py [hunk 18]
```diff
@@ -0,0 +1,501 @@
...
+                    batch_log = 'Train [{0}][{1}/{2}]\t' \
+                                'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t' \
+                                'Data Time {data_time.val:.3f} ({data_time.avg:.3f})\t' \
+                                'Loss {losses.val:.4f} ({losses.avg:.4f})\t' \
+                                'Top-1 acc {top1.val:.3f} ({top1.avg:.3f})\t' \
+                                'Top-5 acc {top5.val:.3f} ({top5.avg:.3f})\tlr {lr:.5f}'. \
+                        format(epoch + 1, i, nBatch - 1, batch_time=batch_time, data_time=data_time,
+                               losses=losses, top1=top1, top5=top5, lr=lr)
+                    logger.info(batch_log)
+            # TODO: print current network architecture
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/trainer.py]: ???

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 19]
```diff
@@ -0,0 +1,471 @@
...
+                                [_i for _i in range(sample + 1, len(mutable.choices))]
+            self.log_prob = torch.log(probs[sample])
+            self.current_prob_over_ops = probs
+            self.AP_path_wb.data[sample] = 1.0
+        # avoid over-regularization
+        for choice in mutable.choices:
+            for _, param in choice.named_parameters():
+                param.grad = None
+
+    def _delta_ij(self, i, j):
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: could be a class static function

File: examples/nas/proxylessnas/ops.py [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+        out, idx = self.mobile_inverted_conv(x)
+        # TODO: unify idx format
+        if not isinstance(idx, int):
+            idx = (idx == 1).nonzero()
+        if self.op_candidates_list[idx].is_zero_layer():
+            res = x
+        elif self.shortcut is None:
+            res = out
+        else:
+           conv_x = out
```
Comment (ultmaster) [inline examples/nas/proxylessnas/ops.py]: wrong indentation

File: examples/nas/proxylessnas/putils.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    if min_val is None:
+        min_val = divisor
+    new_v = max(min_val, int(v + divisor / 2) // divisor * divisor)
+    # Make sure that round down does not go down by more than 10%.
+    if new_v < 0.9 * v:
+        new_v += divisor
+    return new_v
+
+class AverageMeter(object):
```
Comment (ultmaster) [inline examples/nas/proxylessnas/putils.py]: You can use AverageMeter from nni.nas.pytorch.utils

File: examples/nas/proxylessnas/retrain.py [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+    n_classes = pred.size(1)
+    # convert to one-hot
+    target = torch.unsqueeze(target, 1)
+    soft_target = torch.zeros_like(pred)
+    soft_target.scatter_(1, target, 1)
+    # label smoothing
+    soft_target = soft_target * (1 - label_smoothing) + label_smoothing / n_classes
+    return torch.mean(torch.sum(- soft_target * logsoftmax(pred), 1))
+
+def accuracy(output, target, topk=(1,)):
```
Comment (ultmaster) [inline examples/nas/proxylessnas/retrain.py]: If SDK has provided accuracy, it's better to import directly from SDK.

File: docs/en_US/NAS/Proxylessnas.md [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+                              ckpt_path=args.checkpoint_path,
+                              arch_path=args.arch_path)
+trainer.train()
+trainer.export(args.arch_path)
+```
+The complete example code can be found [here](https://github.com/microsoft/nni/tree/master/examples/nas/proxylessnas).
+
+**Input arguments of ProxylessNasTrainer**
+
+* **model** (*PyTorch model, required*) - The model that users want to tune/search. It has mutables to specify search space.
```
Comment (ultmaster) [inline docs/en_US/NAS/Proxylessnas.md]: I believe these are docstrings. Provide a link to the reference page would be better.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 24]
```diff
@@ -0,0 +1,471 @@
...
+    forward_mode = None
+    def __init__(self, mutable):
+        """
+        Parameters
+        ----------
+        mutable : LayerChoice
+            A LayerChoice in user model
+        """
+        super(MixedOp, self).__init__()
+        self.AP_path_alpha = nn.Parameter(torch.Tensor(mutable.length))
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: suggest `ap_path_alpha` (lowercase) to make linter happy.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 25]
```diff
@@ -0,0 +1,471 @@
...
+        dict
+            the choice of each mutable, i.e., LayerChoice
+        """
+        result = dict()
+        for mutable in self.undedup_mutables:
+            assert isinstance(mutable, LayerChoice)
+            index, _ = mutable.registered_module.chosen_index
+            # pylint: disable=not-callable
+            result[mutable.key] = F.one_hot(torch.tensor(index), num_classes=mutable.length).view(-1).bool()
+        return result
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: missing endline.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py [hunk 26]
```diff
@@ -0,0 +1,471 @@
...
+        ----------
+        model : pytorch model
+            The model that users want to tune, it includes search space defined with nni nas apis
+        """
+        super(ProxylessNasMutator, self).__init__(model)
+        self._unused_modules = None
+        self.mutable_list = []
+        for mutable in self.undedup_mutables:
+            self.mutable_list.append(mutable)
+            mutable.registered_module = MixedOp(mutable)
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: If I'm right, when users try to dump a checkpoint of the model, they will get two copies of weight? One as `xxx.choices.xxx`, another as `registered_module.xxx`.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/utils.py [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+
+    _, pred = output.topk(maxk, 1, True, True)
+    pred = pred.t()
+    correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+    res = []
+    for k in topk:
+        correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
+        res.append(correct_k.mul_(100.0 / batch_size))
+    return res
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/utils.py]: missing endline.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/utils.py [hunk 28]
```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import torch
+from torch import nn as nn
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/pytorch/proxylessnas/utils.py]: import torch.nn as nn

File: docs/en_US/NAS/Proxylessnas.md (hunk 23 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: a simple link is not user friendly, let's keep it for now

File: examples/nas/proxylessnas/datasets.py (hunk 1 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/datasets.py]: updated thanks

File: examples/nas/proxylessnas/datasets.py (hunk 3 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/datasets.py]: good point, let's do it in a dedicated pr

File: examples/nas/proxylessnas/main.py (hunk 4 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/main.py]: good catch, fixed.

File: examples/nas/proxylessnas/model.py (hunk 7 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/model.py]: removed, thanks.

File: examples/nas/proxylessnas/ops.py (hunk 20 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: good catch, thanks.

File: examples/nas/proxylessnas/ops.py (hunk 9 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: this is a code snippet in the official repo

File: examples/nas/proxylessnas/ops.py (hunk 10 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: good point, will update

File: examples/nas/proxylessnas/ops.py (hunk 11 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: what kind of reference? paper?

File: examples/nas/proxylessnas/putils.py (hunk 13 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/putils.py]: good catch, will remove them.

File: examples/nas/proxylessnas/putils.py (hunk 21 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/putils.py]: yes, removed.

File: examples/nas/proxylessnas/retrain.py (hunk 22 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/retrain.py]: for future refactor

File: src/sdk/pynni/nni/nas/pytorch/fixed.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/fixed.py]: you can do this refactor. I simply removed the commented line.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 16 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: done

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 24 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: done

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 17 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: which part do you mean? 'two'? or forward_mode?

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 19 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: updated, thanks

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 26 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: not two copies, `xxx.choices.xxx` stores candiate ops, `registered_module.xxx` stores architecture weights.

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/trainer.py (hunk 18 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/trainer.py]: removed

File: examples/nas/proxylessnas/ops.py (hunk 9 above)
Comment (xuehui1991) [reply examples/nas/proxylessnas/ops.py]: what's that means?

File: examples/nas/proxylessnas/ops.py (hunk 11 above)
Comment (xuehui1991) [reply examples/nas/proxylessnas/ops.py]: the paper/wiki/any other link that introduces this layer

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 17 above)
Comment (xuehui1991) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: what're the "two" forward_mode means?  what's kind of forward_mode are "two" forward_mode?

File: docs/en_US/NAS/Overview.md [hunk 29]
```diff
@@ -93,6 +94,26 @@ cd ../darts
 python3 retrain.py --arc-checkpoint ../pdarts/checkpoints/epoch_2.json
 ```
 
+### ProxylessNAS
+
+The paper [ProxylessNAS: Direct Neural Architecture Search on Target Task and Hardware](https://arxiv.org/pdf/1812.00332.pdf) removes proxy, it directly learn the architectures for large-scale target tasks and target hardware platforms. It addresses high memory consumption issue of differentiable NAS and reduces the computational cost to the same level of regular training while still allowing a large candidate set.
```
Comment (xuehui1991) [inline docs/en_US/NAS/Overview.md]: `learn` ---> ` learns`??

Comment (xuehui1991) [inline docs/en_US/NAS/Overview.md]: `It addresses high memory consumption` --> `It addresses the high memory consumption`

File: examples/nas/proxylessnas/ops.py (hunk 9 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: I don't know.

File: docs/en_US/NAS/Proxylessnas.md [hunk 30]
```diff
@@ -0,0 +1,63 @@
+# ProxylessNAS on NNI
+
+## Introduction
+
+The paper [ProxylessNAS: Direct Neural Architecture Search on Target Task and Hardware](https://arxiv.org/pdf/1812.00332.pdf) removes proxy, it directly learn the architectures for large-scale target tasks and target hardware platforms. They address high memory consumption issue of differentiable NAS and reduce the computational cost to the same level of regular training while still allowing a large candidate set. Please refer to the paper for the details.
```
Comment (xuehui1991) [inline docs/en_US/NAS/Proxylessnas.md]: same typo here.

File: docs/en_US/NAS/Proxylessnas.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+* **warmup_epochs** (*int, optional, default = 25*) - The number of epochs to do during warmup.
+* **arch_valid_frequency** (*int, optional, default = 1*) - The frequency of printing validation result.
+* **load_ckpt** (*bool, optional, default = False*) - Whether to load checkpoint.
+* **ckpt_path** (*str, optional, default = None*) - checkpoint path, if load_ckpt is True, ckpt_path cannot be None.
+* **arch_path** (*str, optional, default = None*) - The path to store chosen architecture.
+
+
+## Implementation
+
+The implementation on NNI is based on the [offical implementation](https://github.com/mit-han-lab/ProxylessNAS). The offical implementation supports two training approaches: gradient descent and RL based, and support different targeted hardwared, including 'mobile', 'cpu', 'gpu8', 'flops'. In our current implementation on NNI, gradient descent training approach is supported, but has not supported different hardwares. The complete support is ongoing.
```
Comment (xuehui1991) [inline docs/en_US/NAS/Proxylessnas.md]: hardwared?

Comment (xuehui1991) [inline docs/en_US/NAS/Proxylessnas.md]: offical?

File: docs/en_US/NAS/Proxylessnas.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+
+Below we will describe implementation details. Like other one-shot NAS algorithms on NNI, ProxylessNAS is composed of two parts: *search space* and *training approach*. For users to flexibily define their own search space and use built-in ProxylessNAS training approach, we put the specified search space in [example code](https://github.com/microsoft/nni/tree/master/examples/nas/proxylessnas) using [NNI NAS interface](NasInterface.md), and put the training approach in [SDK](https://github.com/microsoft/nni/tree/master/src/sdk/pynni/nni/nas/pytorch/proxylessnas).
```
Comment (xuehui1991) [inline docs/en_US/NAS/Proxylessnas.md]: flexibily ?

File: docs/en_US/NAS/Proxylessnas.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+
+![](../../img/proxylessnas.png)
+
+ProxylessNAS training approach is composed of ProxylessNasMutator and ProxylessNasTrainer. ProxylessNasMutator instantiates MixedOp for each mutable (i.e., LayerChoice), and manage architecture weights in MixedOp. **For DataParallel**, architecture weights should be included in user model. Specifically, in ProxylessNAS implementation, we add MixedOp to the corresponding mutable (i.e., LayerChoice) as a member variable. The mutator also exposes two member functions, i.e., `arch_requires_grad`, `arch_disable_grad`, for the trainer to control the training of architecture weights.
+
+ProxylessNasMutator also implements the forward logic of the mutables (i.e., LayerChoice).
+
+## Reproduce Results
+
+Ongoing...
```
Comment (xuehui1991) [inline docs/en_US/NAS/Proxylessnas.md]: report the number for now? I think this is the thing that user cares about.

File: examples/nas/proxylessnas/ops.py (hunk 11 above)
Comment (QuanluZhang) [reply examples/nas/proxylessnas/ops.py]: added

File: src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py (hunk 17 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/proxylessnas/mutator.py]: added the following in `forward` docstring:

The 'two' mode is explained in section 3.2.1 in the paper.
The 'full_v2' mode is explained in Appendix D in the paper.

File: docs/en_US/NAS/Overview.md (hunk 29 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Overview.md]: fixed

Comment (QuanluZhang) [reply docs/en_US/NAS/Overview.md]: I think remove `the` is better

File: docs/en_US/NAS/Proxylessnas.md (hunk 30 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: fixed, thanks

File: docs/en_US/NAS/Proxylessnas.md (hunk 31 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: fixed, thanks

File: docs/en_US/NAS/Proxylessnas.md (hunk 32 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: fixed, thanks

File: docs/en_US/NAS/Proxylessnas.md (hunk 31 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: fixed, thanks

File: docs/en_US/NAS/Proxylessnas.md (hunk 33 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: yes, still ongoing.

Comment (xuehui1991) [reply docs/en_US/NAS/Proxylessnas.md]: I think we should put the number we got for now before we merge into master.

Comment (QuanluZhang) [reply docs/en_US/NAS/Proxylessnas.md]: I only need one day or two to get the results, let's merge it first, after all the numbers are get, I will post it.

**Pull request merged**